### PR TITLE
Update PrimitiveFieldType$STRING.java for UTF-8

### PR DESCRIPTION
**Pull request by Theosakamg**
Add support of UTF-8!
For ~i18n (eg. TTS).
### Docs
1. ["A string must always contain UTF-8 encoded or 7-bit ASCII text."](https://developers.google.com/protocol-buffers/docs/proto#scalar)
2. ["unicode strings are currently not supported as a ROS data type. utf-8 should be used to be compatible with ROS string serialization. "](http://wiki.ros.org/msg)
### Quote

It works. 
- testing with python node.
- for bug test, send string msg with ["éè"](http://www.utf8-chartable.de/unicode-utf8-table.pl?start=128&number=128&utf8=0x&unicodeinhtml=hex), on receive no Multibyte-char managed.
- check for best implement, just a fix!

### CODE REVIEW

**Comment by stonier** [comment]
@damonkohler will this have any unforeseen side effects?

**Comment by damonkohler** [comment]
LGTM but please also fix the serialize method.

File: message_generation/src/main/java/org/ros/internal/message/field/PrimitiveFieldType.java [hunk 1]
```diff
@@ -581,7 +581,7 @@ public int getSerializedSize() {
     public String deserialize(ChannelBuffer buffer) {
       int length = buffer.readInt();
       ByteBuffer stringBuffer = buffer.readSlice(length).toByteBuffer();
-      return Charset.forName("US-ASCII").decode(stringBuffer).toString();
+      return Charset.forName("UTF-8").decode(stringBuffer).toString();
```
Comment (damonkohler) [inline message_generation/src/main/java/org/ros/internal/message/field/PrimitiveFieldType.java]: You also need to specify "UTF8" to getBytes() on line 574. See http://docs.oracle.com/javase/tutorial/i18n/text/string.html

Comment (Theosakamg) [inline message_generation/src/main/java/org/ros/internal/message/field/PrimitiveFieldType.java]: my apologies...
I have not tested the serializer in every context.
I use ROS on Android, which [by default is UTF-8](http://developer.android.com/reference/java/nio/charset/Charset.html#defaultCharset%28%29)

File: message_generation/src/main/java/org/ros/internal/message/field/PrimitiveFieldType.java [hunk 2]
```diff
@@ -679,6 +680,7 @@ public String getJavaTypeName() {
   };
 
   private static final ImmutableSet<String> TYPE_NAMES;
+  private static final Charset DEFAULT_CHARSET = Charset.forName("utf-8");
```
Comment (damonkohler) [inline message_generation/src/main/java/org/ros/internal/message/field/PrimitiveFieldType.java]: Please use "UTF-8"

**Comment by damonkohler** [comment]
Last nit and then LG.

**Comment by Theosakamg** [comment]
@stonier : I testing the change with common case. Do you have other ideas for test cases ?

in org.ros.internal.message.MessageSerializationTest

``` java
  @Test
  public void testStringUTF8() {
    std_msgs.String message = defaultMessageFactory.newFromType(std_msgs.String._TYPE);
    message.setData("éêè €àáßëœ 文字化け");
    checkSerializeAndDeserialize(message);

    // i18n test case
    // base on http://www.inter-locale.com/whitepaper/learn/learn-to-test.html

    // Combining Marks and Accents test
    message.setData("àéîōũ");
    checkSerializeAndDeserialize(message);

    // DOS 860 test
    message.setData("você nós mãe avô irmã criança");
    checkSerializeAndDeserialize(message);

    // Windows-1252 test
    message.setData("€ŒœŠš™©‰ƒ");
    checkSerializeAndDeserialize(message);

    // Turkish test
    message.setData("ışık bir İyi Günler");
    checkSerializeAndDeserialize(message);

    // Dakuten and handakuten marks test
    message.setData("がざばだぱか゛さ゛た゛は");
    checkSerializeAndDeserialize(message);

    // Combining Grapheme Joiner character
    message.setData("אִ͏ַ");
    checkSerializeAndDeserialize(message);

    // Bidi with Latin test
    message.setData("abcאבגדabc ");
    checkSerializeAndDeserialize(message);

    message.setData("אבגדabcאבגד");
    checkSerializeAndDeserialize(message);

    message.setData("אבגד012אבגד");
    checkSerializeAndDeserialize(message);

    message.setData("אבגד 012 012");
    checkSerializeAndDeserialize(message);

    // Complex Scripts test
    message.setData("สวัสดี");
    checkSerializeAndDeserialize(message);

    message.setData("டாஹ்கோ");
    checkSerializeAndDeserialize(message);

    message.setData("بِسْمِ اللّهِ الرَّحْمـَنِ الرَّحِيمِ");
    checkSerializeAndDeserialize(message);

    // Numeric Shaping test
    message.setData("عدد مارس ١٩٩٨");
    checkSerializeAndDeserialize(message);

    // Common Scripts and Encodings test
    message.setData("Слава Жанна Ювеналий Ярополк");
    checkSerializeAndDeserialize(message);
  }
```

**Comment by damonkohler** [comment]
Can you please add those tests to this change?

**Comment by Theosakamg** [comment]
I am not sure of the location. (I use the groovy branch.)

**Comment by damonkohler** [comment]
Sorry, I don't understand the problem. Can you please add the message serialization tests you pasted above to this patch?

**Comment by Theosakamg** [comment]
The problem was : I use groovy stack and the test is available in https://github.com/rosjava/rosjava_core/blob/master/rosjava/src/test/java/org/ros/internal/message/MessageSerializationTest.java (always available in master)
But with the news stack is in https://github.com/rosjava/rosjava_bootstrap/blob/master/message_generation/src/test/java/org/ros/internal/message/RawMessageSerializationTest.java
(If I understand)

It's done in rosjava_core and rosjava_bootstrap repository. It's ok for you?

**Comment by Theosakamg** [comment]
Apologies I just see the difference (Raw vs Std_msg)

File: message_generation/src/main/java/org/ros/internal/message/field/PrimitiveFieldType.java [hunk 3]
```diff
@@ -679,6 +680,7 @@ public String getJavaTypeName() {
   };
 
   private static final ImmutableSet<String> TYPE_NAMES;
+  private static final Charset DEFAULT_CHARSET = Charset.forName("UTF-8");
```
Comment (damonkohler) [inline message_generation/src/main/java/org/ros/internal/message/field/PrimitiveFieldType.java]: Please move this to the top of the class (line 34).

**Comment by damonkohler** [comment]
Looks good, thanks! Just one last nit. The constant DEFAULT_CHARSET belongs at the top of the class.

**Comment by Theosakamg** [comment]
Done.

**Comment by stonier** [comment]
@mh0rst is this the error you were also seeing?

http://pastebin.com/LV31s8g0

This was coming from a ros answers post: http://answers.ros.org/question/140371/compilation-problem-in-rosjava-master-branch/

**Comment by mh0rst** [comment]
@stonier yes, exactly. The fix is trivial, I created a pull request if that helps.

**Comment by stonier** [comment]
Definitely. @damonkohler will come undoubtedly across it when he's free.

**Pull request merged**